### ISSUE TITLE
RuntimeError: Dimension out of range (expected to be in range of [-1, 0], but got 1)

### ISSUE THREAD

**Issue by alioz1967** [rusty1s/pytorch_geometric#1446]
I've initiated the attributes as follows, can't figure out what is wrong or if I've initiated the data attributes correctly. Could you please advise how I should debug this?

```python
data=Data(x=x,edge_index=edge_index.contiguous())
edge_attr = torch.ones(data.edge_index.size(0))#keep this as dummy var for now
data.train_mask=train_mask
data.test_mask=test_mask
data.y=y

class Net(torch.nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = SplineConv(42, 16, dim=1, kernel_size=2)
        self.conv2 = SplineConv(16, 3, dim=1, kernel_size=2)

    def forward(self):
        x, edge_index, edge_attr = data.x, data.edge_index, data.edge_attr
        x = F.dropout(x, training=self.training)
        x = F.elu(self.conv1(x, edge_index, edge_attr))
        x = F.dropout(x, training=self.training)
        x = self.conv2(x, edge_index, edge_attr)
        return F.log_softmax(x, dim=1)

def train():
    model.train()
    optimizer.zero_grad()
    F.nll_loss(model()[data.train_mask], data.y[data.train_mask]).backward()
    optimizer.step()
for epoch in range(1, 201):
    train()
    #set_trace()
    log = 'Epoch: {:03d}, Train: {:.4f}, Test: {:.4f}'
    print(log.format(epoch, *test()))
```

Traceback (most recent call last):
  File "C:/Users/aoz/Documents/c/g/nf-cpu_for_Debug.py", line 385, in train
    F.nll_loss(model()[data.train_mask], data.y[data.train_mask]).backward()
  File "C:\Users\aoz\AppData\Local\Continuum\anaconda3\envs\python3\lib\site-packages\torch\nn\modules\module.py", line 550, in __call__
    result = self.forward(*input, **kwargs)
  File "C:/Users/aoz/Documents/c/g/nf-cpu_for_Debug.py", line 366, in forward
    x = F.elu(self.conv1(x, edge_index, edge_attr))
  File "C:\Users\aoz\AppData\Local\Continuum\anaconda3\envs\python3\lib\site-packages\torch\nn\modules\module.py", line 550, in __call__
    result = self.forward(*input, **kwargs)
  File "C:\Users\aoz\AppData\Local\Continuum\anaconda3\envs\python3\lib\site-packages\torch_geometric\nn\conv\spline_conv.py", line 122, in forward
    out = self.propagate(edge_index, x=x, edge_attr=edge_attr, size=size)
  File "C:\Users\aoz\AppData\Local\Continuum\anaconda3\envs\python3\lib\site-packages\torch_geometric\nn\conv\message_passing.py", line 236, in propagate
    out = self.message(**msg_kwargs)
  File "C:\Users\aoz\AppData\Local\Continuum\anaconda3\envs\python3\lib\site-packages\torch_geometric\nn\conv\spline_conv.py", line 135, in message
    self.degree)
RuntimeError: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript (most recent call last):
  File "C:\Users\aoz\AppData\Local\Continuum\anaconda3\envs\python3\lib\site-packages\torch_spline_conv\basis.py", line 10, in spline_basis
                 is_open_spline: torch.Tensor,
                 degree: int) -> Tuple[torch.Tensor, torch.Tensor]:
    return torch.ops.torch_spline_conv.spline_basis(pseudo, kernel_size,
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
                                                    is_open_spline, degree)
RuntimeError: Dimension out of range (expected to be in range of [-1, 0], but got 1)

**Comment by rusty1s** [comment]
Fixable via
```python
edge_attr = torch.ones((data.edge_index.size(0), 1))
```

**Comment by alioz1967** [comment]
Thank you for the quick reply! that cleared the error but now getting this 
Traceback (most recent call last):
  File "C:/Users/aoz/Documents/c/g/nf-cpu_for_Debug.py", line 399, in <module>
    train()
  File "C:/Users/aoz/Documents/c/g/nf-cpu_for_Debug.py", line 371, in train
    F.nll_loss(model()[data.train_mask], data.y[data.train_mask]).backward()
  File "C:\Users\aoz\AppData\Local\Continuum\anaconda3\envs\python3\lib\site-packages\torch\nn\modules\module.py", line 550, in __call__
    result = self.forward(*input, **kwargs)
  File "C:/Users/aoz/Documents/c/g/nf-cpu_for_Debug.py", line 356, in forward
    x = F.elu(self.conv1(x, edge_index, edge_attr))
  File "C:\Users\aoz\AppData\Local\Continuum\anaconda3\envs\python3\lib\site-packages\torch\nn\modules\module.py", line 550, in __call__
    result = self.forward(*input, **kwargs)
  File "C:\Users\aoz\AppData\Local\Continuum\anaconda3\envs\python3\lib\site-packages\torch_geometric\nn\conv\spline_conv.py", line 122, in forward
    out = self.propagate(edge_index, x=x, edge_attr=edge_attr, size=size)
  File "C:\Users\aoz\AppData\Local\Continuum\anaconda3\envs\python3\lib\site-packages\torch_geometric\nn\conv\message_passing.py", line 236, in propagate
    out = self.message(**msg_kwargs)
  File "C:\Users\aoz\AppData\Local\Continuum\anaconda3\envs\python3\lib\site-packages\torch_geometric\nn\conv\spline_conv.py", line 136, in message
    return spline_weighting(x_j, self.weight, *data)
RuntimeError: Caught an unknown exception!

**Comment by rusty1s** [comment]
Hard to say what may have caused this. Can you show me an example to reproduce this?

**Comment by alioz1967** [comment]
I have included your previous fix for `edge_attr` below and additional print statements in case it's useful to give an idea. (in case it's not enough I'm trying to figure out how I can share something reproducible, without the actual dataset)
```
data=Data(x=x,edge_index=edge_index.contiguous())
edge_attr = torch.ones((data.edge_index.size(0), 1))#keep this as dummy var for now
data.train_mask=train_mask
data.test_mask=test_mask
data.y=y
data.edge_attr=edge_attr
class Net(torch.nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = SplineConv(42, 16, dim=1, kernel_size=2)
        self.conv2 = SplineConv(16, 3, dim=1, kernel_size=2)

    def forward(self):
        x, edge_index, edge_attr = data.x, data.edge_index, data.edge_attr
        x = F.dropout(x, training=self.training)
        x = F.elu(self.conv1(x, edge_index, edge_attr))
        x = F.dropout(x, training=self.training)
        x = self.conv2(x, edge_index, edge_attr)
        return F.log_softmax(x, dim=1)

def train():
    model.train()
    optimizer.zero_grad()
    F.nll_loss(model()[data.train_mask], data.y[data.train_mask]).backward()
    optimizer.step()
for epoch in range(1, 201):
    train()
    #set_trace()
    log = 'Epoch: {:03d}, Train: {:.4f}, Test: {:.4f}'
    print(log.format(epoch, *test()))
print("type data.y, size data.y, data.y.type",type(data.y),data.y.size(),data.y.type())
print(" type of data(train_mask, test_mask)",type(data('train_mask', 'test_mask')))
print("type data.trainmask, type data.testmask)",type(data.train_mask), type(data.test_mask))
print("trainmask size", data.train_mask.size())
print("testmask size",data.test_mask.size())
print("type data.x, size data.x, data.x.type",type(data.x),data.x.size(),data.x.type())
print("type data.edge_index, size data.edge_index, data.edge_index.type",type(data.edge_index),data.edge_index.size(),data.edge_index.type())
print("type data.edge_attr, size data.edge_attr, data.edge_attr.type", type(data.edge_attr), data.edge_attr.size(), data.edge_attr.type())

type data.y, size data.y, data.y.type <class 'torch.Tensor'> torch.Size([121793]) torch.LongTensor
 type of data (trainmask, test mask) <class 'generator'>
type data.trainmask, type data.testmask) <class 'torch.Tensor'> <class 'torch.Tensor'>
trainmask size torch.Size([121793])
testmask size torch.Size([121793])
type data.x, size data.x, data.x.type <class 'torch.Tensor'> torch.Size([121793, 42]) torch.FloatTensor
type data.edge_index, size data.edge_index, data.edge_index.type <class 'torch.Tensor'> torch.Size([2, 121793]) torch.LongTensor
type data.edge_attr, size data.edge_attr, data.edge_attr.type <class 'torch.Tensor'> torch.Size([2, 1]) torch.FloatTensor
Traceback (most recent call last):
  File "C:/Users/aoz/Documents/c/g/nf-cpu_for_Debug.py", line 396, in <module>
    train()
  File "C:/Users/aoz/Documents/c/g/nf-cpu_for_Debug.py", line 371, in train
    F.nll_loss(model()[data.train_mask], data.y[data.train_mask]).backward()
  File "C:\Users\aoz\AppData\Local\Continuum\anaconda3\envs\python3\lib\site-packages\torch\nn\modules\module.py", line 550, in __call__
    result = self.forward(*input, **kwargs)
  File "C:/Users/aoz/Documents/c/g/nf-cpu_for_Debug.py", line 356, in forward
    x = F.elu(self.conv1(x, edge_index, edge_attr))
  File "C:\Users\aoz\AppData\Local\Continuum\anaconda3\envs\python3\lib\site-packages\torch\nn\modules\module.py", line 550, in __call__
    result = self.forward(*input, **kwargs)
  File "C:\Users\aoz\AppData\Local\Continuum\anaconda3\envs\python3\lib\site-packages\torch_geometric\nn\conv\spline_conv.py", line 122, in forward
    out = self.propagate(edge_index, x=x, edge_attr=edge_attr, size=size)
  File "C:\Users\aoz\AppData\Local\Continuum\anaconda3\envs\python3\lib\site-packages\torch_geometric\nn\conv\message_passing.py", line 236, in propagate
    out = self.message(**msg_kwargs)
  File "C:\Users\aoz\AppData\Local\Continuum\anaconda3\envs\python3\lib\site-packages\torch_geometric\nn\conv\spline_conv.py", line 136, in message
    return spline_weighting(x_j, self.weight, *data)
RuntimeError: Caught an unknown exception!

Process finished with exit code 1```

**Comment by alioz1967** [comment]
It seems to be fixed with `edge_attr = torch.ones((data.edge_index.size(1), 1))`  instead of .size(0) Thank you.

**Comment by rusty1s** [comment]
Ah yes, of course :)

**Comment by LiuCMU** [comment]
Hi, I run into a similar error:
IndexError: **Dimension out of range (expected to be in range of [-1, 0], but got -2).**

My situation is similar: I have implemented my own dataset. The Data object and dataset object functions well. But I encounter this error when pass the data into a network.

I have sent you invitation to my repo so that you can reproduce the error, if you are interested:-) 

Here is some screeshots:
![image](https://user-images.githubusercontent.com/60156077/88224123-8632d500-cc36-11ea-96bd-8f92abff0f15.png)
![image](https://user-images.githubusercontent.com/60156077/88224179-99de3b80-cc36-11ea-9a39-c3bc976fc868.png)
![image](https://user-images.githubusercontent.com/60156077/88224213-a8c4ee00-cc36-11ea-8ac8-389984635a0d.png)
![image](https://user-images.githubusercontent.com/60156077/88224283-bed2ae80-cc36-11ea-88fc-fc3e9c8246c7.png)
![image](https://user-images.githubusercontent.com/60156077/88224316-cd20ca80-cc36-11ea-9c55-b1244fbef825.png)

**Comment by rusty1s** [comment]
You need to reshape your node features to shape [num_nodes, num_features]. In case your features are one-dimensional, please do `data.x = data.x.view(-1, 1)`.

**Comment by LiuCMU** [comment]
Thank you! I transformed the node featuers to shape [num_nodes, 1] by adding `data.x = data.x.view(-1, 1)`. But that  raised another error. (sorry if that was a stupid error, I am still trying to improve myself in thie new area).
![image](https://user-images.githubusercontent.com/60156077/88234513-095c2700-cc47-11ea-8af8-33d18138f5fa.png)

**Comment by rusty1s** [comment]
I see. Node features need to be of type float/double:
`data.x = data.x.to(torch.float)`

**Comment by LiuCMU** [comment]
Thank you so much!
I checked other data types and made sure that `data.x` is` torch.foloat`, `data.edge_index` is` torch.long`. I'm not sure with `data.y`, I put it as `torch.long` since I am doing a binary classificaiton task. This solved the fomer errors. 

But I found the following error. It seems to be the weight. I didn't have weights in my Data object.
![image](https://user-images.githubusercontent.com/60156077/88299176-14a76500-ccd0-11ea-9e03-da738f28775d.png)
